File: internal/scaffold/scaffold_test.go
package scaffold

import (
	"context"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"reporting-db-migrations/internal/db"
	"reporting-db-migrations/internal/fs"
	"reporting-db-migrations/internal/types"
)

func TestEnsureTransitionFiles_NoBlockedTables(t *testing.T) {
	s := New()
	layout := fs.Layout{
		Objects: []*fs.Object{{
			NormalizedKey: "r/views/v1",
			Kind:          "views",
		}},
	}
	plan := &types.MigrationPlan{
		Objects: []types.PlannedObject{{
			ObjectRef:     types.ObjectRef{NormalizedKey: "r/views/v1"},
			PlannedAction: types.ActionCreateObject,
		}},
	}
	columns := map[string][]db.TableColumn{}

	created, err := s.EnsureTransitionFiles(context.Background(), types.Config{}, layout, plan, columns)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if created {
		t.Error("expected created=false for non-blocked plan")
	}
}

func TestEnsureTransitionFiles_BlockedTableCreatesScaffold(t *testing.T) {
	s := New()
	baseDir := t.TempDir()
	layout := fs.Layout{
		Objects: []*fs.Object{{
			NormalizedKey: "r/tables/t1",
			Kind:          "tables",
			ObjectName:    "t1",
			SchemaName:    "r",
		}},
	}
	plan := &types.MigrationPlan{
		Blocked: true,
		Objects: []types.PlannedObject{{
			ObjectRef:     types.ObjectRef{NormalizedKey: "r/tables/t1", Kind: "tables", ObjectName: "t1", SchemaName: "r"},
			PlannedAction: types.ActionReprocessChangedBlocked,
		}},
	}
	columns := map[string][]db.TableColumn{
		"r/tables/t1": {
			{Name: "id", TypeName: "int", Nullable: false},
			{Name: "name", TypeName: "nvarchar", Nullable: false},
		},
	}
	cfg := types.Config{SQLBase: baseDir}

	created, err := s.EnsureTransitionFiles(context.Background(), cfg, layout, plan, columns)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !created {
		t.Fatal("expected created=true for blocked table")
	}

	dir := filepath.Join(baseDir, "r", "tables", "_migrations", "t1")
	entries, err := os.ReadDir(dir)
	if err != nil {
		t.Fatalf("readdir: %v", err)
	}
	var files []string
	for _, e := range entries {
		if strings.HasSuffix(e.Name(), ".sql") {
			files = append(files, e.Name())
		}
	}
	if len(files) == 0 {
		t.Fatal("no scaffold file created")
	}

	fullPath := filepath.Join(dir, files[0])
	data, err := os.ReadFile(fullPath)
	if err != nil {
		t.Fatalf("failed to read generated file: %v", err)
	}
	content := string(data)
	if !strings.Contains(content, "-- rmig: transition-scaffold") {
		t.Error("scaffold content missing transition-scaffold marker")
	}
	if !strings.Contains(content, "[r].[t1]") {
		t.Error("scaffold content missing table reference")
	}
}

func TestEnsureTransitionFiles_Idempotent(t *testing.T) {
	s := New()
	baseDir := t.TempDir()
	layout := fs.Layout{
		Objects: []*fs.Object{{
			NormalizedKey: "r/tables/t1",
			Kind:          "tables",
			ObjectName:    "t1",
			SchemaName:    "r",
		}},
	}
	plan := &types.MigrationPlan{
		Blocked: true,
		Objects: []types.PlannedObject{{
			ObjectRef:     types.ObjectRef{NormalizedKey: "r/tables/t1", Kind: "tables", ObjectName: "t1", SchemaName: "r"},
			PlannedAction: types.ActionReprocessChangedBlocked,
		}},
	}
	columns := map[string][]db.TableColumn{}
	cfg := types.Config{SQLBase: baseDir}

	created1, err := s.EnsureTransitionFiles(context.Background(), cfg, layout, plan, columns)
	if err != nil {
		t.Fatalf("first call error: %v", err)
	}
	created2, err := s.EnsureTransitionFiles(context.Background(), cfg, layout, plan, columns)
	if err != nil {
		t.Fatalf("second call error: %v", err)
	}
	if !created1 {
		t.Fatal("first call: expected created=true")
	}
	if created2 {
		t.Error("second call: expected created=false (idempotent)")
	}
}

func TestEnsureTransitionFiles_ExistingRealTransitionSkips(t *testing.T) {
	s := New()
	baseDir := t.TempDir()

	targetDir := filepath.Join(baseDir, "r", "tables", "_migrations", "t1")
	if err := os.MkdirAll(targetDir, 0755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	os.WriteFile(filepath.Join(targetDir, "001_abc_add_col.sql"), []byte("ALTER TABLE t1 ADD col INT;\n"), 0644)

	layout := fs.Layout{
		Objects: []*fs.Object{{
			NormalizedKey: "r/tables/t1",
			Kind:          "tables",
			ObjectName:    "t1",
			SchemaName:    "r",
		}},
	}
	plan := &types.MigrationPlan{
		Blocked: true,
		Objects: []types.PlannedObject{{
			ObjectRef:     types.ObjectRef{NormalizedKey: "r/tables/t1", Kind: "tables", ObjectName: "t1", SchemaName: "r"},
			PlannedAction: types.ActionReprocessChangedBlocked,
		}},
	}
	columns := map[string][]db.TableColumn{}
	cfg := types.Config{SQLBase: baseDir}

	created, err := s.EnsureTransitionFiles(context.Background(), cfg, layout, plan, columns)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if created {
		t.Error("expected created=false (real transition exists)")
	}
}

func TestEnsureTransitionFiles_AutoAddColumn(t *testing.T) {
	s := New()
	baseDir := t.TempDir()

	sqlPath := filepath.Join(baseDir, "r", "tables", "t1.sql")
	if err := os.MkdirAll(filepath.Dir(sqlPath), 0755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	createTableSQL := "CREATE TABLE [r].[t1] (\n  [id] INT NOT NULL,\n  [new_col] NVARCHAR(100) NULL\n)"
	if err := os.WriteFile(sqlPath, []byte(createTableSQL), 0644); err != nil {
		t.Fatalf("writefile: %v", err)
	}

	layout := fs.Layout{
		Objects: []*fs.Object{{
			CachedFile:    fs.CachedFile{AbsPath: sqlPath},
			Path:          "r/tables/t1.sql",
			NormalizedKey: "r/tables/t1",
			Kind:          "tables",
			ObjectName:    "t1",
			SchemaName:    "r",
		}},
	}
	plan := &types.MigrationPlan{
		Blocked: true,
		Objects: []types.PlannedObject{{
			ObjectRef:     types.ObjectRef{NormalizedKey: "r/tables/t1", Kind: "tables", ObjectName: "t1", SchemaName: "r"},
			PlannedAction: types.ActionReprocessChangedBlocked,
		}},
	}

	dbColumns := map[string][]db.TableColumn{
		"r/tables/t1": {
			{Name: "id", TypeName: "INT", Nullable: false},
		},
	}

	cfg := types.Config{SQLBase: baseDir}
	created, err := s.EnsureTransitionFiles(context.Background(), cfg, layout, plan, dbColumns)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !created {
		t.Fatal("expected created=true for auto ADD COLUMN")
	}

	dir := filepath.Join(baseDir, "r", "tables", "_migrations", "t1")
	entries, _ := os.ReadDir(dir)
	var files []string
	for _, e := range entries {
		if strings.HasSuffix(e.Name(), ".sql") {
			files = append(files, e.Name())
		}
	}
	if len(files) == 0 {
		t.Fatal("no migration file created")
	}

	fullPath := filepath.Join(dir, files[0])
	data, err := os.ReadFile(fullPath)
	if err != nil {
		t.Fatalf("readfile: %v", err)
	}
	content := string(data)
	if strings.Contains(content, "-- rmig: transition-scaffold") {
		t.Error("auto migration should NOT be a scaffold")
	}
	if !strings.Contains(content, "ALTER TABLE") {
		t.Error("auto migration should contain ALTER TABLE")
	}
	if !strings.Contains(content, "new_col") {
		t.Error("auto migration should reference new column")
	}
}

func TestEnsureTransitionFiles_AutoAddColumn_FallsBackToScaffoldOnDrop(t *testing.T) {
	s := New()
	baseDir := t.TempDir()

	sqlPath := filepath.Join(baseDir, "r", "tables", "t1.sql")
	if err := os.MkdirAll(filepath.Dir(sqlPath), 0755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	createTableSQL := "CREATE TABLE [r].[t1] (\n  [new_col] INT NOT NULL\n)" // "id" column dropped
	if err := os.WriteFile(sqlPath, []byte(createTableSQL), 0644); err != nil {
		t.Fatalf("writefile: %v", err)
	}

	layout := fs.Layout{
		Objects: []*fs.Object{{
			CachedFile:    fs.CachedFile{AbsPath: sqlPath},
			Path:          "r/tables/t1.sql",
			NormalizedKey: "r/tables/t1",
			Kind:          "tables",
			ObjectName:    "t1",
			SchemaName:    "r",
		}},
	}
	plan := &types.MigrationPlan{
		Blocked: true,
		Objects: []types.PlannedObject{{
			ObjectRef:     types.ObjectRef{NormalizedKey: "r/tables/t1", Kind: "tables", ObjectName: "t1", SchemaName: "r"},
			PlannedAction: types.ActionReprocessChangedBlocked,
		}},
	}

	dbColumns := map[string][]db.TableColumn{
		"r/tables/t1": {
			{Name: "id", TypeName: "INT", Nullable: false},
			{Name: "new_col", TypeName: "INT", Nullable: false},
		},
	}

	cfg := types.Config{SQLBase: baseDir}
	created, err := s.EnsureTransitionFiles(context.Background(), cfg, layout, plan, dbColumns)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !created {
		t.Fatal("expected created=true, fallback to scaffold")
	}

	dir := filepath.Join(baseDir, "r", "tables", "_migrations", "t1")
	entries, _ := os.ReadDir(dir)
	for _, e := range entries {
		if strings.HasSuffix(e.Name(), ".sql") {
			data, _ := os.ReadFile(filepath.Join(dir, e.Name()))
			if strings.Contains(string(data), "-- rmig: transition-scaffold") {
				return
			}
		}
	}
	t.Error("expected scaffold fallback when columns are dropped")
}

File: crates/core/tests/scaffold_test.rs
use std::collections::HashMap;

use migrator_core::config::Config;
use migrator_core::db::TableColumn;
use migrator_core::domain::{
    share, Action, ObjectEntry, ObjectKey, Script, ScriptKey, ScriptKind, Workspace,
};
use migrator_core::export::{MigrationPlan, PlannedObject};
use migrator_core::scaffold;

#[test]
fn blocked_table_creates_scaffold_file() {
    let base = tempfile::tempdir().unwrap();
    let mut cfg = Config::default();
    cfg.sql_base = base.path().to_string_lossy().into();
    cfg.database = "dactests".into();
    let plan = MigrationPlan {
        blocked: true,
        objects: vec![PlannedObject {
            normalized_key: "r/tables/t1".into(),
            object_path: "r/tables/t1.sql".into(),
            schema_name: "r".into(),
            kind: "tables".into(),
            object_name: "t1".into(),
            planned_action: Action::ReprocessChangedBlocked,
            exists: true,
            checksum: [0; 32],
            database_name: share("dactests"),
            parent_name: Default::default(),
            git: None,
            transition_paths: Vec::new(),
        }],
        ..Default::default()
    };
    let mut cols = HashMap::new();
    cols.insert(
        "r/tables/t1".into(),
        vec![
            TableColumn {
                name: "id".into(),
                type_name: "int".into(),
                nullable: false,
            },
            TableColumn {
                name: "name".into(),
                type_name: "nvarchar".into(),
                nullable: false,
            },
        ],
    );
    let created = scaffold::ensure(&cfg, &Workspace::default(), &plan, &cols).unwrap();
    assert!(created);
    let dir = base.path().join("dactests/r/tables/_migrations/t1");
    let content = std::fs::read_dir(&dir)
        .unwrap()
        .flatten()
        .find(|e| e.path().extension().map(|x| x == "sql").unwrap_or(false))
        .map(|e| std::fs::read_to_string(e.path()).unwrap())
        .unwrap();
    assert!(content.contains("-- rmig: transition-scaffold"));
    assert!(content.contains("[r].[t1]"));
}

#[test]
fn auto_add_column_when_file_has_new_col() {
    let base = tempfile::tempdir().unwrap();
    let mut cfg = Config::default();
    cfg.sql_base = base.path().to_string_lossy().into();
    cfg.database = "dactests".into();
    let sql_path = base.path().join("r/tables/t1.sql");
    std::fs::create_dir_all(sql_path.parent().unwrap()).unwrap();
    std::fs::write(
        &sql_path,
        "CREATE TABLE [r].[t1] (\n  [id] INT NOT NULL,\n  [new_col] NVARCHAR(100) NULL\n)",
    )
    .unwrap();
    let sk = ScriptKey::from_path("r/tables/t1.sql");
    let mut ws = Workspace::default();
    let script_id = ws.insert_script(Script {
        key: sk.clone(),
        kind: ScriptKind::Object,
        abs_path: migrator_core::domain::share(sql_path.to_string_lossy().as_ref()),
        checksum: None,
    });
    let db_id = ws.intern_database(share("dactests"));
    ws.adopt_dense_entries(vec![ObjectEntry::with_staging_key(
        ObjectKey::new("r", "tables", "t1"),
        script_id,
        [0; 32],
        false,
        db_id,
    )]);
    let plan = MigrationPlan {
        blocked: true,
        objects: vec![PlannedObject {
            normalized_key: "r/tables/t1".into(),
            object_path: "r/tables/t1.sql".into(),
            schema_name: "r".into(),
            kind: "tables".into(),
            object_name: "t1".into(),
            planned_action: Action::ReprocessChangedBlocked,
            exists: true,
            checksum: [0; 32],
            database_name: share("dactests"),
            parent_name: Default::default(),
            git: None,
            transition_paths: Vec::new(),
        }],
        ..Default::default()
    };
    let mut cols = HashMap::new();
    cols.insert(
        "r/tables/t1".into(),
        vec![TableColumn {
            name: "id".into(),
            type_name: "int".into(),
            nullable: false,
        }],
    );
    assert!(scaffold::ensure(&cfg, &ws, &plan, &cols).unwrap());
    let dir = base.path().join("dactests/r/tables/_migrations/t1");
    let content = std::fs::read_dir(&dir)
        .unwrap()
        .flatten()
        .find(|e| e.path().extension().map(|x| x == "sql").unwrap_or(false))
        .map(|e| std::fs::read_to_string(e.path()).unwrap())
        .unwrap();
    assert!(!content.contains("-- rmig: transition-scaffold"));
    assert!(content.contains("ALTER TABLE"));
    assert!(content.contains("new_col"));
}

#[test]
fn auto_add_column_declines_when_table_drops_existing_column_edge_case() {
    let content = scaffold_for_table(
        "CREATE TABLE [r].[t1] (\n  [new_col] INT NULL\n)",
        vec![TableColumn {
            name: "id".into(),
            type_name: "int".into(),
            nullable: false,
        }],
    );

    assert!(content.contains("-- rmig: transition-scaffold"));
    assert!(!content.contains("ALTER TABLE"));
}

#[test]
fn auto_add_column_declines_unsafe_computed_column_regression() {
    let content = scaffold_for_table(
        "CREATE TABLE [r].[t1] (\n  [id] INT NOT NULL,\n  [calc] AS ([id] + 1)\n)",
        vec![TableColumn {
            name: "id".into(),
            type_name: "int".into(),
            nullable: false,
        }],
    );

    assert!(content.contains("-- rmig: transition-scaffold"));
    assert!(!content.contains("ALTER TABLE"));
}

#[test]
fn auto_add_column_declines_unbracketed_computed_column_regression() {
    let content = scaffold_for_table(
        "CREATE TABLE [r].[t1] (\n  [id] INT NOT NULL,\n  calc AS id + 1\n)",
        vec![TableColumn {
            name: "id".into(),
            type_name: "int".into(),
            nullable: false,
        }],
    );

    assert!(content.contains("-- rmig: transition-scaffold"));
    assert!(!content.contains("ALTER TABLE"));
}

fn scaffold_for_table(table_sql: &str, db_columns: Vec<TableColumn>) -> String {
    let base = tempfile::tempdir().unwrap();
    let mut cfg = Config::default();
    cfg.sql_base = base.path().to_string_lossy().into();
    cfg.database = "dactests".into();
    let sql_path = base.path().join("r/tables/t1.sql");
    std::fs::create_dir_all(sql_path.parent().unwrap()).unwrap();
    std::fs::write(&sql_path, table_sql).unwrap();
    let sk = ScriptKey::from_path("r/tables/t1.sql");
    let mut ws = Workspace::default();
    let script_id = ws.insert_script(Script {
        key: sk,
        kind: ScriptKind::Object,
        abs_path: migrator_core::domain::share(sql_path.to_string_lossy().as_ref()),
        checksum: None,
    });
    let db_id = ws.intern_database(share("dactests"));
    ws.adopt_dense_entries(vec![ObjectEntry::with_staging_key(
        ObjectKey::new("r", "tables", "t1"),
        script_id,
        [0; 32],
        false,
        db_id,
    )]);
    let plan = MigrationPlan {
        blocked: true,
        objects: vec![PlannedObject {
            normalized_key: "r/tables/t1".into(),
            object_path: "r/tables/t1.sql".into(),
            schema_name: "r".into(),
            kind: "tables".into(),
            object_name: "t1".into(),
            planned_action: Action::ReprocessChangedBlocked,
            exists: true,
            checksum: [0; 32],
            database_name: share("dactests"),
            parent_name: Default::default(),
            git: None,
            transition_paths: Vec::new(),
        }],
        ..Default::default()
    };
    let mut cols = HashMap::new();
    cols.insert("r/tables/t1".into(), db_columns);
    assert!(scaffold::ensure(&cfg, &ws, &plan, &cols).unwrap());
    let dir = base.path().join("dactests/r/tables/_migrations/t1");
    std::fs::read_dir(&dir)
        .unwrap()
        .flatten()
        .find(|e| e.path().extension().map(|x| x == "sql").unwrap_or(false))
        .map(|e| std::fs::read_to_string(e.path()).unwrap())
        .unwrap()
}
